README.md
# gimp-stable-diffusion (replicate.com fork)

This repository includes a GIMP plugin for communication with the  stable-diffusion model on replicate.com.

It is a fork of <https://github.com/blueturtleai> which provides plugins backed by a Colb notbook, or by <https://stablehorde.net/>. Hopefully at some point a single great plugin will just support many different backends.

Please check HISTORY.md for the latest changes.

## Installation
### Download files

To download the files of this repository click on "Code" and select "Download ZIP". In the ZIP you will find the file "gimp-stable-diffusion.py". This is the code for the GIMP plugin. You don't need the other files in the ZIP.

### GIMP

The plugin is tested in GIMP 2.10 and runs most likely in all 2.* releases. Excluded is 2.99, because it's already Python 3 based.

1. Start GIMP and open the preferences dialog via edit/preferences and scroll down to "folders". Expand "folders" and click on "plug-ins". Select the folder which includes your username and copy the path.

2. Open the file explorer, navigate to this directory and copy the file "gimp-stable-diffusion.py" from the repository into this directory. If you are on MacOS or Linux, change the file permissions to 755.

3. Restart GIMP. You should now see the new menu "AI". If you don't see this, something went wrong. Please check in this case "Troubleshooting/GIMP" for possible solutions. The menu has one item "Replicate.com img2img". This item can't currently be selected. This only works, when you opened an image before.

## Generate images
Now we are ready for generating images.

1. Start GIMP and open an image or create a new one. It is recommended, that the image size is not larger than 512x512 as the model has been trained on this size. If you want to have larger images, use an external upscaler instead. The generated image will have the dimensions of the init image. But it may be resized to make sure, that the dimensions are a multiple of 64. The larger the image, the longer it takes to generate it and the more GPU ressources and RAM is used. If it is too larger, you will run out of memory.

2. Select the new AI/Stable im2img menu item. A dialog will open, where you can enter the details for the image generation.

   - **Use init image:** If you want to seed the process with the current image or start from scratch (like with text2img).

     TODO: Support proper inpainting with a mask.

   - **Prompt Strength:** Prompt strength when using init image. 1.0 corresponds to full destruction of information in init image. 0.8 is a good value to use.

   - **Guidance Scale:** Scale for classifier-free guidance. 7.5 is a good value to use.

   - **Steps:** How many steps the AI should use to generate the image. The higher the value, the more the AI will work on details. But it also means, the longer the generation takes and the more expensive the run is. 50 is a good value to use.

   - **Seed:** This parameter is optional. If it is empty, a random seed will be generated on the server. If you use a seed, the same image is generated again in the case the same parameters for init strength, steps, etc. are used. A slightly different image will be generated, if the parameters are modified.

     TODO: Include the seed somehow in the output (layer name?)

   - **Number of images:** Number of images, which are created in one run. The more images you create, the more server ressources will be used and the longer you have to wait until the generated images are displayed in GIMP.

   - **Prompt:** How the generated image should look like.

   - **Replicate.com token:** Insert the replicate.com token found on <https://replicate.com/account>. Do not share this with others (e.g. in screenshots)

3. Click on the OK button. The values you inserted into the dialog and the init image will be transmitted to the server, which starts now generating the image. When the image has been generated successfully, it will be shown as a new image in GIMP.

<!--

not done yet

### Inpainting 
Inpainting means replacing a part of an existing image. For example if you don't like the face on an image, you can replace it. **Inpainting is currently still in experimental stage. So, please don't expect perfect results.** The experimental stage is caused by the server side and not by GIMP.

For inpainting it's necessary to prepare the input image because the AI needs to know which part you want to replace. For this purpose you replace this image part by transparency. To do so, open the init image in GIMP and select "Layer/Transparency/Add alpha channel". Select now the part of the image which should be replaced and delete it. You can also use the eraser tool. 

For the prompt you use now a description of the new image. For example the image shows currently "a little girl running over a meadow with a balloon" and you want to replace the balloon by a parachute. You just write now "a little girl running over a meadow with a parachute".

-->

## Troubleshooting
### GIMP
#### AI menu is not shown
##### Linux
   - If you get this error ```gimp: LibGimpBase-WARNING: gimp: gimp_wire_read(): error```, it's very likely, that you have a GIMP version installed, which doesn't include Python. Check, if you have got the menu "Filters > Python-Fu > Console". If it is missing, please install GIMP from here: https://flathub.org/apps/details/org.gimp.GIMP.

##### macOS
   - Please double check, if the permissions of the plugin py file are set to 755. It seems, that changing permissions doesn't work via the file manager. Please open a terminal, cd to the plugins directory and run "chmod ugo+x *py".
   
##### macOS/Linux
   - Open a terminal an try to run the plugin py file manually via ```python <path-to-plugin-folder>/gimp-stable-diffusion.py```. You should see the error message, that "gimpfu" is unknown. Make sure, that you are running Python 2, as this version is used by GIMP. If other errors occur, please reinstall GIMP.

## FAQ

**Will GIMP 3 be supported?**
Yes, the plugin will be ported to GIMP 3.

**How do I report an error or request a new feature?** Please open a new issue in this repository.




HISTORY.md
# History
## GIMP Plugin

### Unrelesed

#### Changes

- Port to replicate.com

### 1.1.2
#### Changes
- Exception handling further improved

### 1.1.1
#### Changes
- Data is now transferred to the server using https

### 1.1.0
#### Changes
- Inpainting is now supported

### 1.0.2
#### Changes
- **Image count**: It is now possible to create up to 4 images in one run

- **Seed**: Can now be blank if not used (before -1)

- **Backend URL**: Needs no longer to end with "/"

- **Exception handling**: Has been improved

### 1.0.1
#### CHANGES
- **API**: Data is now transferred as JSON. It is checked, if the API version of the client and the server matches. If mismatch, user is requested to update plugin and/or server. 

- **Seed**: The seed of the generated image is now displayed in an additional layer. It can be passed to the server to generate the same image again. This only works, if the same parameters (init strength, number of steps, etc) are used.

### 1.0.0
Initial version

## Stable-Diffusion server
### 1.2.0
#### Changes
- It's no longer necessary to upload model files yourself to your Google drive. If a model file is missing, it is automatically downloaded from Huggingface. A Huggingface account and acceptance of the terms of service is still needed.

### 1.1.2
#### Changes
- Instead of ngrok now cloudflare is used as the tunnel provider. The switch is done because several users reported problems. Additional benefit is, that it's no longer necessary to register on ngrok and enter the authkey. Thanks for suggesting this @opencoca.

### 1.1.1
#### Changes
- Mounting Google drive and setting model path are now two separated steps

### 1.1.0
#### Changes
- The notebook is now based on the deforum notebook (before pharmapsychotic)
- Inpainting is now supported

### 1.0.3
#### Changes
- **Image count**: It is now possible to create up to 4 images in one run

### 1.0.2
#### Changes
- **API**: Data is now transferred in JSON. It is checked, if API version of the client and server matches. If mismatch, code 405 is returned.

- **Seed**: If a seed != -1 is transferred to the server, this seed is used for image generation.

### 1.0.1
#### Changes
- **Image dimensions**: The image dimensions are now no longer fixed to 512x512. The dimensions of the init image are now used instead. The dimensions are adjusted to a multiple of 64.

### 1.0.0
Initial version


gimp-stable-diffusion.py
#!/usr/bin/python

# v1.1.2

import urllib2
import tempfile
import os
import base64
import json
import re
import ssl
import sched, time

from gimpfu import *

# https://replicate.com/stability-ai/stable-diffusion/versions/a9758cbfbd5f3c2094457d996681af52552901775aa2d6dd0b17fd15df959bef
MODEL_ID = "a9758cbfbd5f3c2094457d996681af52552901775aa2d6dd0b17fd15df959bef"

API_URL = "https://api.replicate.com/v1/"

INIT_FILE = "init.png"
GENERATED_FILE = "generated.png"

# HEADER_ACCEPT = "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8"
HEADER_USER_AGENT = "nomeata/gimp-stable-diffusion"

# check every 5 seconds
CHECK_WAIT = 5
checkMax = None
s = sched.scheduler(time.time, time.sleep)


initFile = r"{}".format(os.path.join(tempfile.gettempdir(), INIT_FILE))
generatedFile = r"{}".format(os.path.join(tempfile.gettempdir(), GENERATED_FILE))

def getImageData(image, drawable):
   pdb.file_png_save_defaults(image, drawable, initFile, initFile)
   initImage = open(initFile, "rb")
   encoded = base64.b64encode(initImage.read())
   return encoded

def img2img(image, drawable, useInit, guidanceScale, promptStrength, steps, seed, imageCount, prompt, token):
   data = {
      "prompt": prompt,
      "width": int(image.width),
      "height": int(image.height),
      "prompt_strength": float(promptStrength),
      "num_outputs": int(imageCount),
      "num_inference_steps": int(steps),
      "guidance_scale": float(guidanceScale),
   }

   if seed:
     data.update({"seed": int(seed)})

   if useInit:
     imageData = getImageData(image, drawable)
     data.update({"init_image": "data:image/png;base64," + imageData})

   request = {
     "version": MODEL_ID,
     "input": data
   }

   url = API_URL + "predictions"
   req_body = json.dumps(request)
   headers = {
     "Authorization": "Token " + token,
     "Content-Type": "application/json",
     "User-agent": HEADER_USER_AGENT
   }

   request = urllib2.Request(url=url, data=req_body, headers=headers)
   pdb.gimp_progress_set_text("Starting dreaming now...")

   response = urllib2.urlopen(request)
   handleStatus(response, token)

def handleStatus(response, token):
   data = response.read()
   #print(data)
   data = json.loads(data)

   pdb.gimp_progress_set_text("Status: " + data["status"])

   if data["status"] == "starting" or data["status"] == "processing":
      s.enter(CHECK_WAIT, 1, checkStatus, (data["urls"]["get"],token))
      s.run()
   elif data["status"] == "failed" or data["status"] == "canceled":
      pdb.gimp_message(data["error"])
   elif data["status"] == "succeeded":
       for image_url in data["output"]:
          pdb.gimp_progress_set_text("Downloading image")

          headers = {
            "Authorization": "Token " + token,
            "User-agent": HEADER_USER_AGENT
          }
          request = urllib2.Request(url=image_url, headers=headers)
          response = urllib2.urlopen(request)
          with open(generatedFile, mode="wb") as d:
            d.write(response.read())

          imageLoaded = pdb.file_png_load(generatedFile, generatedFile)
          pdb.gimp_display_new(imageLoaded)

          # The seed is only part of the log, so tricky to parse

          # # image, drawable, x, y, text, border, antialias, size, size_type, fontname
          # pdb.gimp_text_fontname(imageLoaded, None, 2, 2, str(image["seed"]), -1, TRUE, 12, 1, "Sans")
          #pdb.gimp_image_set_active_layer(imageLoaded, imageLoaded.layers[1])


       if os.path.exists(initFile):
            os.remove(initFile)

       if os.path.exists(generatedFile):
            os.remove(generatedFile)

def checkStatus(url, token):
  headers = {
    "Authorization": "Token " + token,
    "User-agent": HEADER_USER_AGENT
  }
  request = urllib2.Request(url=url, headers=headers)
  response = urllib2.urlopen(request)
  handleStatus(response, token)

register(
   "replicate_img2img",
   "replicate_img2img",
   "replicate_img2img",
   "BlueTurtleAI, nomeata",
   "BlueTurtleAI, nomeata",
   "2022",
   "<Image>/AI/Replicate.com img2img",
   "*",
   [
      (PF_TOGGLE, "useInit", "Use init image", False),
      #(PF_SLIDER, "maskBrightness", "Inpainting\nMask Brightness", 1.0, (0.0, 1.0, 0.1)),
      #(PF_SLIDER, "maskContrast", "Inpainting\nMask Contrast", 1.0, (0.0, 1.0, 0.1)),
      (PF_SLIDER, "guidanceScale", "Guidance Scale", 7.5, (0, 20, 0.5)),
      (PF_SLIDER, "promptStrength", "Prompt strength", 0.8, (0.0, 1.0, 0.1)),
      (PF_SLIDER, "steps", "Steps", 20, (10, 150, 1)),
      (PF_STRING, "seed", "Seed (optional)", ""),
      (PF_SLIDER, "imageCount", "Number of images", 1, (1, 4,1)),
      (PF_STRING, "prompt", "Prompt", ""),
      (PF_STRING, "token", "replicate.com token", "")
   ],
   [],
   img2img
)

main()
